### ISSUE TITLE
[material] Simplify building responsive layouts based on window size classes.

### ISSUE THREAD

**Issue by fischerscode** [flutter/flutter#142343]
### Use case

When building modern applications, making them responsive is essential. It would be great if there would be an easy way to build different layouts based on the [window size class](https://m3.material.io/foundations/layout/applying-layout/window-size-classes#2bb70e22-d09b-4b73-9c9f-9ef60311ccc8).

This could also be a foundation for issues like #142310.

### Proposal

In `material`, there should be an enum `WindowSizeClass` that represents the different [window size classes](https://m3.material.io/foundations/layout/applying-layout/window-size-classes#2bb70e22-d09b-4b73-9c9f-9ef60311ccc8). Additionally, there should be a `StatelessWidget` called `ResponsiveBuilder`, that uses its `ResponsiveWidgetBuilder` to build the child based on the current `WindowSizeClass`.

Since "window size classes" are a part of Material 3, this should not be done in a new package.

**Comment by YukiAttano** [comment]
I would like to suggest that these is implemented in the MediaQueryData to allow MediaQuery.heightSizeClassOf() and  MediaQuery.widthSizeClassOf().

I've seen that your PR https://github.com/flutter/flutter/pull/142348 was closed.
Instead of the hardcoded values, i would go with the official documentation https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes?hl=en that only states three classes.

This would also enable Flutter developers to create a user experience that user are familiar with.

I think that your Builder Function needs more context as people might misunderstand its behavior. 
If it act like the OrientationBuilder, which gives you the orientation of the currently available space for the current widget, it would mostly be useless as the only widgets that require this information are toplevel Widgets like screens/pages. (Most widgets would be < 600p big and the builder would never be called)

If it just builds the current widgets with the overall window size class, it might provide a widget with information that might lead to misbehavior and hard to debug layouts. 
Consider the following situation: 
Instead of developing a widget that just relies on its current space, it now changes its layout based on the overall size while its own space is more limited than you would expect in the situation of writing.

**Comment by kerberjg** [comment]
@YukiAttano I disagree this should be part of MediaQuery, as it's platform-specific.

[Material has its own size classes](https://m3.material.io/foundations/layout/applying-layout/window-size-classes), while [Cupertino/iOS has its own](https://m3.material.io/foundations/layout/applying-layout/window-size-classes), with different breakpoints, classifications etc.

I suggest this effort be split into platform-specific issues.

**Comment by YukiAttano** [comment]
> @YukiAttano I disagree this should be part of MediaQuery, as it's platform-specific.
> 
> [Material has its own size classes](https://m3.material.io/foundations/layout/applying-layout/window-size-classes), while [Cupertino/iOS has its own](https://m3.material.io/foundations/layout/applying-layout/window-size-classes), with different breakpoints, classifications etc.
> 
> I suggest this effort be split into platform-specific issues.

That's a good point. But i would say it is less platform specific and more 'design' specific if Material and Cupertino have different definitions and not Android/iOS/etc.

In that case, i would like to see it in the Flutter library as it is somehow defined in the official docs.
But i do also understand if it is never implemented as this 'enforces' constraints not everyone want to follow.
I will create a package in the next week that allows such behavior. Could you send me the link to the Apple docs for their guidelines? 
(I normally ignore the Cupertino design as i don't like it/don't like Apple's Marketing/don't like their Docs and the endless links to other subpages)

**Comment by kerberjg** [comment]
Fair enough, I was trying to find a better word to define them, perhaps "design implementations/libraries"?

Here's the link to the iOS docs, I realized I pasted in the wrong one in the comment: https://developer.apple.com/design/human-interface-guidelines/layout#Device-size-classes

I actually already wrote all of the necessary code for the Cupertino side of things as I mainly work with iOS, I'd be very happy to share!

**Comment by YukiAttano** [comment]
> Fair enough, I was trying to find a better word to define them, perhaps "design implementations/libraries"?
> 
> Here's the link to the iOS docs, I realized I pasted in the wrong one in the comment: https://developer.apple.com/design/human-interface-guidelines/layout#Device-size-classes
> 
> I actually already wrote all of the necessary code for the Cupertino side of things as I mainly work with iOS, I'd be very happy to share!

I have the code part of Material ready to go, sounds like we are a good match 😀.
But i want to redesign it to make it non-fixed on the guidelines for everyone.

Seeing the apple docs lets me remember why i ignored that part.
I will be back from vacation in a few days and send you a link to the then published package. If it meets your expectations i will be happy to have you on board.

**Comment by kerberjg** [comment]
Wonderful, keep me posted! 🤗

I do have a rudimentary solution for Material as well, but I'm curious to see what you came up with

**Comment by YukiAttano** [comment]
@kerberjg 
https://pub.dev/packages/responsive_ux
I have sent you an invitation and apologies to all the people who get spammed by our chat.

To have an easier communication, i will reach out to you with my private mail (just so you know in case it ends up in your spam)

**Comment by YukiAttano** [comment]
> I am adding this code and assigning copyright to Google,
> 
> `// Directly in the Material Design 3 spec we need 
> /// WindowSize as measured by width and margin and 
> /// middle spacing. Indirectly, we also need the indirectly 
> /// implied textDimens as evidenced by the discussion of 
> /// information density per small and large screens.
> ///
> /// The SizedContext extension pairs with this enum, see
> /// [SizedContext].
> ///
> /// Material Design Spec docs start at Window 
> /// Size Classes see
> /// https://m3.material.io/foundations/layout/applying-layout/window-size-classes
> ///
> /// @author Fredrick Allan Grott
> enum WindowSize {
>   
>   compact(0, 599, 16, 0, 4),
>   medium(600,839 , 24, 24, 8),
>   expanded(840 , 1199 , 24, 24,16 ),
>   large(1200 ,1599 ,24 ,24, 32  ),
>   extraLarge(1600 , double.infinity , 24, 24, 64);
> 
> 
>   const WindowSize(this.beginWidthRange, this.endWidthRange, this.margin, this.middleSpacing, this.dimens);
> 
>   final double beginWidthRange;
> 
>   final double endWidthRange;
> 
>   final double margin;
> 
>   final double middleSpacing;
> 
>   // for padding insets, its 
>   // obliquely mentioned when talking about 
>   // alignment needing to change for large screen
>   // content items in large screens.
>   final double dimens;
> 
> }
> 
> extensions AppContext on BuildContext {
>   
>   MediaQueryData get mq => MediaQuery.of(this);
>   
>   /// Returns an appWindowSize that contains the beginWidthRange, endWidthRange, margin, 
>   /// middleSpacing, and dimens
>   WindowSize get appWindowSize {
>     if (mq.size.width >= WindowSize.compact.endWidthRange) {
>       return WindowSize.compact;
>     } else if (mq.size.width >= WindowSize.medium.endWidthRange) {
>       return WindowSize.medium;
>     } else if (mq.size.width >= WindowSize.expanded.endWidthRange) {
>       return WindowSize.expanded;
>     } else if (mq.size.width >= WindowSize.large.endWidthRange) {
>       return WindowSize.large;
>     } else {
>       return WindowSize.extraLarge;
>     }
>   }
>   
> }
> 
> `
> 
> Note handled both the direct width ranges, margins, and spacing and added dimens for the padding spacing via inserts for big screens.
> 
> If Google requires my further sig of copyright let me know via this issue an my comment
> 
> I added to speed things up as Apple has announced iphone foldable for 2026 and we need this 
> complete so that the issues with flutter adaptive scaffold can be further addressed.
> 
> All of its accessible via the BuildContext which is a better solution than some other ones...
> 
> I did borrow the idea from GSkinner's own SizedContext package...
> 

That's a nice idea, i don't know if you posted everything or just a short example but don't forget the height WindowSize.

I am not sure where i've read that, but someone said that Flutter don't want a static implementation of the WindowSizes (but they defined some in Material specs...)

Feel free to check my linked package, maybe it is of help for you.

**Comment by martijn00** [comment]
FYI: https://github.com/flutter/packages/tree/main/packages/flutter_adaptive_scaffold implements most of this. Maybe it's an idea to team up to shape the API and later on get it merged into material?

**Comment by fischerscode** [comment]
If needed, I'd happily assist from September.

**Comment by YukiAttano** [comment]
> FYI: https://github.com/flutter/packages/tree/main/packages/flutter_adaptive_scaffold implements most of this. Maybe it's an idea to team up to shape the API and later on get it merged into material?
> 
> I'm working with @gspencergoog on new proposals for the API.

Thats a great idea. I checked the repo and think the implementation got a lot more hardcoded values than necessary.
Especially this [array](https://github.com/flutter/packages/blob/b9715b7eac7d97428c4acde94a75dad5a7525668/packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart#L91) seem to me that it breaks the pure purpose of the whole idea.

If you guys are up to chat, is there is a better place to share our thoughts or shall we stay in this Issue?

**Comment by martijn00** [comment]
This is the draft doc i'm working in: https://docs.google.com/document/d/1DMFfGrHv_V1Je6_wrfyu0iY3ngJgn317OscfNbuRMRA/pub

It would be great to get some help and feedback shaping the api for that.

I'm not sure the array is a problem in the current structure. In compose they just support 3 panes and depending on some checks you get different navigation or more panes. Flutter uses a more complex SlotLayout but that also means you got to support a lot more customisations.

I'm at Flutter Discord as `Martijn00` with @gspencergoog if you want to discuss things.

**Comment by martijn00** [comment]
@fischerscode do you have time to help out improving adaptive_scaffold to meet the other requirements?

It would be great to split up some of the work and getting towards a stable base that could later on be merged into flutter itself.
